pkg/planner/core/operator/logicalop/logical_datasource.go

```diff
@@ -621,3 +633,90 @@ func preferKeyColumnFromTable(dataSource *DataSource, originColumns []*expressio
 	}
 	return resultColumn, resultColumnInfo
 }
+
+// analyzeFTSFunc checks whether FTS function is used and is a valid one.
+// Then convert the function to index call because it can not be executed without the index.
+func (ds *DataSource) analyzeFTSFunc() error {
+	if !ds.SCtx().HasFTSFunc() {
+		return nil
+	}
+	idx2FastCheck := make(map[*model.IndexInfo]intset.FastIntSet)
+	idSetForCheck := intset.NewFastIntSet()
+	for _, index := range ds.FtsIndexes {
+		s := intset.NewFastIntSet()
+		for _, col := range index.Columns {
+			s.Insert(int(ds.TableInfo.Columns[col.Offset].ID))
+		}
+		idx2FastCheck[index] = s
+	}
+	var matchedIdx *model.IndexInfo
+	var matchedFunc *expression.ScalarFunction
+	var matchedCondPos int
+	matchedColumns := make([]*expression.Column, 0, 2)
+	for i, cond := range ds.PushedDownConds {
+		sf, ok := cond.(*expression.ScalarFunction)
+		if !ok || sf.FuncName.L != ast.FTSMatchWord {
+			if expression.ContainsFullTextSearchFn(cond) {
+				return plannererrors.ErrWrongUsage.FastGen(plannererrors.FTSWrongPlace)
+			}
+			continue
+		}
+		idSetForCheck.Clear()
+		for i := 1; i < len(sf.GetArgs()); i++ {
+			col := sf.GetArgs()[i].(*expression.Column)
+			idSetForCheck.Insert(int(col.ID))
+		}
+		var currentIndex *model.IndexInfo
+		for idx, set := range idx2FastCheck {
+			// The used columns in the FTS function should be a subset of the index columns.
+			if idSetForCheck.SubsetOf(set) {
+				currentIndex = idx
+				break
+			}
+		}
+		// If the index is not found, it means that the FTS function is not valid.
+		if currentIndex == nil {
+			return errors.New("Full text search can only be used with a matching fulltext index and a columnar storage")
+		}
+		// Currently TiDB doesn't support multiple fulltext search functions used with multiple index calls.
+		if matchedIdx != nil {
+			return errors.New("Current TiDB doesn't support multiple fulltext search functions used with multiple index calls")
+		}
+		matchedIdx = currentIndex
+		matchedFunc = sf
+		matchedCondPos = i
+		matchedColumns = matchedColumns[:0]
+		for i := 1; i < len(sf.GetArgs()); i++ {
+			col := sf.GetArgs()[i].(*expression.Column)
+			matchedColumns = append(matchedColumns, col)
+		}
+	}
+	// Fulltext index must be used. So we prune all other possible access paths.
+	ds.PossibleAccessPaths = slices.DeleteFunc(ds.PossibleAccessPaths, func(path *util.AccessPath) bool {
+		return path.StoreType == kv.TiKV
+	})
+	if len(ds.PossibleAccessPaths) == 0 {
+		return plannererrors.ErrWrongUsage.FastGen("Full text search can be only executed in a columnar storage")
```

Comment:
The `SCtx().HasFTSFunc()` check is at the session level.
If we have two tables in a query and only one of them involves FTS functions, then the logic for the other table should not come here.

Reply:
we can improve it later, just like the checking of whether a table has dirty read content.

Reply:
I think the current logic would cause an error when the SQL contains another table without FTS function.